crates/sui-rest-api/src/transactions/resolve/literal.rs

```diff
@@ -0,0 +1,787 @@
+{
+    let n: T = match value {
+        UnresolvedValue::Number(n) => T::try_from(*n).map_err(|e| {
+            RestError::new(
+                axum::http::StatusCode::BAD_REQUEST,
+                format!(
+                    "literal cannot be resolved as {}: {e}",
+                    std::any::type_name::<T>()
+                ),
+            )
+        })?,
+
+        UnresolvedValue::String(s) => s.parse().map_err(|e| {
+            RestError::new(
+                axum::http::StatusCode::BAD_REQUEST,
+                format!(
+                    "literal cannot be resolved as {}: {e}",
+                    std::any::type_name::<T>()
+                ),
+            )
+        })?,
+
+        UnresolvedValue::Null | UnresolvedValue::Bool(_) | UnresolvedValue::Array(_) => {
+            return Err(RestError::new(
+                axum::http::StatusCode::BAD_REQUEST,
+                format!(
+                    "literal cannot be resolved into type {}",
+                    std::any::type_name::<T>()
+                ),
+            ))
+        }
+    };
+
+    bcs::serialize_into(buf, &n)?;
+
+    Ok(())
+}
+
+fn resolve_as_address(buf: &mut Vec<u8>, value: &UnresolvedValue) -> Result<()> {
+    let address = match value {
+        // parse as ObjectID to handle the case where 0x is present or missing
+        UnresolvedValue::String(s) => s.parse::<ObjectID>().map_err(|e| {
+            RestError::new(
+                axum::http::StatusCode::BAD_REQUEST,
+                format!("literal cannot be resolved as bool: {e}"),
+            )
+        })?,
+        UnresolvedValue::Null
+        | UnresolvedValue::Bool(_)
+        | UnresolvedValue::Number(_)
+        | UnresolvedValue::Array(_) => {
+            return Err(RestError::new(
+                axum::http::StatusCode::BAD_REQUEST,
+                "literal cannot be resolved into type address",
+            ))
+        }
+    };
+
+    bcs::serialize_into(buf, &address)?;
+
+    Ok(())
+}
+
+fn resolve_as_string(buf: &mut Vec<u8>, value: &UnresolvedValue) -> Result<()> {
+    match value {
+        UnresolvedValue::String(s) => {
+            bcs::serialize_into(buf, s)?;
+        }
+        UnresolvedValue::Bool(_)
+        | UnresolvedValue::Null
+        | UnresolvedValue::Number(_)
+        | UnresolvedValue::Array(_) => {
+            return Err(RestError::new(
+                axum::http::StatusCode::BAD_REQUEST,
+                "literal cannot be resolved into string",
+            ))
+        }
+    };
+
+    Ok(())
+}
+
+fn resolve_as_option(buf: &mut Vec<u8>, type_: &Type, value: &UnresolvedValue) -> Result<()> {
+    match value {
+        UnresolvedValue::Null => {
+            buf.push(0);
+        }
+        UnresolvedValue::Bool(_)
+        | UnresolvedValue::Number(_)
+        | UnresolvedValue::String(_)
+        | UnresolvedValue::Array(_) => {
+            buf.push(1);
+            resolve_literal_to_type(buf, type_, value)?;
```

Comment:
Maybe I am reading this wrong, but if `resolve_literal_to_type` fails, wouldn't the buffer already have a `1` inserted? Shouldn't this handle the case when the resolve fails to drop the `1` from the buffer?

Reply:
no, if there's an error then the buffer would be thrown away. I could have used the std::io::Write trait instead of a Vec but found this more convenient.

Reply:
I see, then yeah makes sense!